tools/base/fpgad/config_file.c

```diff
@@ -179,7 +183,299 @@ STATIC cfg_plugin_configuration *alloc_configuration(bool enabled,
 
 	return p;
 }
-#endif
+
+STATIC cfg_vendor_device_id *
+cfg_process_plugin_devices(const char *name,
+			   json_object *j_devices)
+{
+	int i;
+	int devs;
+	cfg_vendor_device_id *head = NULL;
+	cfg_vendor_device_id *id = NULL;
+	uint16_t vendor_id;
+	uint16_t device_id;
+	const char *s;
+	char *endptr;
+
+	if (!json_object_is_type(j_devices, json_type_array)) {
+		LOG("'devices' JSON object not array.\n");
+		return NULL;
+	}
+
+	devs = json_object_array_length(j_devices);
+	for (i = 0 ; i < devs ; ++i) {
+		json_object *j_dev = json_object_array_get_idx(j_devices, i);
+		json_object *j_vid;
+		json_object *j_did;
+
+		if (!json_object_is_type(j_dev, json_type_array)) {
+			LOG("%s 'devices' entry %d not array.\n",
+				name, i);
+			goto out_free;
+		}
+
+		if (json_object_array_length(j_dev) != 2) {
+			LOG("%s 'devices' entry %d not array[2].\n",
+				name, i);
+			goto out_free;
+		}
+
+		j_vid = json_object_array_get_idx(j_dev, 0);
+		if (json_object_is_type(j_vid, json_type_string)) {
+			s = json_object_get_string(j_vid);
+			endptr = NULL;
+
+			vendor_id = (uint16_t)strtoul(s, &endptr, 0);
+			if (*endptr != '\0') {
+				LOG("%s malformed Vendor ID at devices[%d]\n",
+					name, i);
+				goto out_free;
+			}
+
+		} else if (json_object_is_type(j_vid, json_type_int)) {
+			vendor_id = (uint16_t)json_object_get_int(j_vid);
+		} else {
+			LOG("%s invalid Vendor ID at devices[%d]\n",
+				name, i);
+			goto out_free;
+		}
+
+		j_did = json_object_array_get_idx(j_dev, 1);
+		if (json_object_is_type(j_did, json_type_string)) {
+			s = json_object_get_string(j_did);
+			endptr = NULL;
+
+			device_id = (uint16_t)strtoul(s, &endptr, 0);
+			if (*endptr != '\0') {
+				LOG("%s malformed Device ID at devices[%d]\n",
+					name, i);
+				goto out_free;
+			}
+
+		} else if (json_object_is_type(j_did, json_type_int)) {
+			device_id = (uint16_t)json_object_get_int(j_did);
+		} else {
+			LOG("%s invalid Device ID at devices[%d]\n",
+				name, i);
+			goto out_free;
+		}
+
+		if (!head) {
+			head = alloc_device(vendor_id, device_id);
+			if (!head) {
+				LOG("malloc failed.\n");
+				goto out_free;
+			}
+
+			id = head;
+		} else {
```

Comment:
Head is initialized to null, would it ever get into the else?

Reply:
This is within the for loop, so !head will only be true the first time through.  Subsequent iterations will it the else case and append to the end of the list via id.

Reply:
👍